### PR TITLE
add algorithm to extract graph roots

### PR DESCRIPTION
**Pull request by sokra**
show only graph roots in chunks by default

**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
yes

**Does this PR introduce a breaking change?**
yes, defaults changed

**What needs to be documented once your changes are merged?**
new `stats.chunkRootModules` option. Shows graph roots for chunks.

### CODE REVIEW

File: lib/Stats.js [hunk 1]
```diff
@@ -788,6 +795,25 @@ class Stats {
 						);
 					}
 				}
+				if (showChunkRootModules) {
+					const rootModules = chunkGraph.getChunkRootModules(chunk);
+					obj.rootModules = rootModules
+						.slice()
+						.sort(sortRealModules)
+						.filter(createModuleFilter("root-of-chunk"))
```
Comment (Connormiha) [inline lib/Stats.js]: We can improve perfarmance
```js
.slice()
.sort(sortRealModules)
.filter(createModuleFilter("root-of-chunk"))
```
=>
```js
.filter(createModuleFilter("root-of-chunk"))
.sort(sortRealModules)
```

File: lib/util/findGraphRoots.js [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+ */
+module.exports = (items, getDependencies) => {
+	const itemToNode = new Map();
+	for (const item of items) {
+		const node = new Node(item);
+		itemToNode.set(item, node);
+	}
+
+	// early exit when there is only a single item
+	if (itemToNode.size <= 1) return items;
```
Comment (Connormiha) [inline lib/util/findGraphRoots.js]: Can we check it in first line of function?
```js
if (items.length <= 1) return items
// or // if (items.size <= 1) return items
```

File: lib/Stats.js (hunk 1 above)
Comment (ooflorent) [reply lib/Stats.js]: No because `createModuleFilter` is impure and requires the modules to be sorted

File: lib/util/findGraphRoots.js (hunk 2 above)
Comment (ooflorent) [reply lib/util/findGraphRoots.js]: It won't catch duplicates

Comment (sokra) [reply lib/util/findGraphRoots.js]: and `items` is an `Iterable` which has no `size` or `length`

**Pull request merged**